examples/resources/mapbox-streets-v6-style.js

```diff
@@ -1,312 +1,674 @@
-      styles[length++] = getIcon(maki);
-    } else if (layer == 'poi_label' && resolution <= 9.554628535647032 &&
-        scalerank == 2 && maki !== 'marker') {
-      styles[length++] = getIcon(maki);
-    } else if (layer == 'poi_label' && resolution <= 4.777314267823516 &&
-        scalerank == 3 && maki !== 'marker') {
-      styles[length++] = getIcon(maki);
-    } else if (layer == 'poi_label' && resolution <= 2.388657133911758 &&
-        scalerank == 4 && maki !== 'marker') {
-      styles[length++] = getIcon(maki);
-    } else if (layer == 'poi_label' && resolution <= 1.194328566955879 &&
-        scalerank >= 5 && maki !== 'marker') {
-      styles[length++] = getIcon(maki);
-    }
-    styles.length = length;
-    return styles;
+function getMapboxStreetsV6Style() {
+  const iconOffsetVariables = {};
+  const iconProperties = {
+    'icon-src': '../data/maki-4.0.0.png',
+    'icon-size': [15, 15],
+    'icon-offset': ['var', ['concat', 'offset-', ['get', 'maki']]],
   };
-}
\ No newline at end of file
+
+  // load sprite info into style variables
+  fetch('../data/maki-4.0.0.json')
+    .then((resp) => resp.json())
+    .then((sprites) => {
+      Object.keys(sprites).forEach((spriteId) => {
+        const offset = sprites[spriteId];
+        const shortId = spriteId.replace(/-[0-9]+$/, '');
+        iconOffsetVariables[`offset-${shortId}`] = [offset.x, offset.y];
+      });
+    });
+
+  /** @type {Array<import('../../src/ol/style/flat.js').Rule>} */
+  return [
+    {
+      filter: [
+        'all',
+        ['==', ['get', 'layer'], 'landuse'],
+        ['==', ['get', 'class'], 'park'],
+      ],
+      style: {
+        'fill-color': '#d8e8c8',
+      },
+    },
+    {
+      filter: [
+        'all',
+        ['==', ['get', 'layer'], 'landuse'],
+        ['==', ['get', 'class'], 'cemetery'],
+      ],
+      style: {
+        'fill-color': '#e0e4dd',
+      },
+    },
+    {
+      filter: [
+        'all',
+        ['==', ['get', 'layer'], 'landuse'],
+        ['==', ['get', 'class'], 'hospital'],
+      ],
+      style: {
+        'fill-color': '#fde',
+      },
+    },
+    {
+      filter: [
+        'all',
+        ['==', ['get', 'layer'], 'landuse'],
+        ['==', ['get', 'class'], 'school'],
+      ],
+      style: {
+        'fill-color': '#f0e8f8',
+      },
+    },
+    {
+      filter: [
+        'all',
+        ['==', ['get', 'layer'], 'landuse'],
+        ['==', ['get', 'class'], 'wood'],
+      ],
+      style: {
+        'fill-color': 'rgb(233,238,223)',
+      },
+    },
+    {
+      filter: ['==', ['get', 'layer'], 'waterway'],
+      style: {
+        'stroke-color': '#a0c8f0',
+        'stroke-width': 1,
+      },
+    },
+    {
+      filter: ['==', ['get', 'layer'], 'water'],
+      style: {
+        'fill-color': '#a0c8f0',
+      },
+    },
+    {
+      filter: [
+        'all',
+        ['==', ['get', 'layer'], 'aeroway'],
+        ['==', ['geometry-type'], 'Polygon'],
```

Comment:
I had been thinking we would add things like `hasFill`, `hasStroke`, and `hasPoint` to the parsing context.  Then we could skip points and lines, for example, in cases where a style only had fill related properties.  I added a similar comment to #15416.

Reply:
That sounds good but the parsing context exists for an expression, not a whole style, so this would have to be done somewhere else. Maybe there could be `hasFill()` etc. utility functions for flat styles?

This filtering would be a great performance improvement, I definitely agree.